### ISSUE TITLE
Trainable assistant answers

### ISSUE THREAD

**Issue by Martovark** [SWE-bench/SWE-smith#98]
Hi!
If I understood correctly, `SWE-agent-LM-32B` was trained on each assistant response (on each agent action).
For example, in this script `SWE-smith/swesmith/train/run/ft_unsloth.py`:
```python
...
trainer = train_on_responses_only(
        trainer,
        instruction_part="<|im_start|>user\n",
        response_part="<|im_start|>assistant\n",
    )
...
```
Have you experimented with not including “bad” actions of the agent in the loss function?
For example, the same repeated actions to correct a piece of code or actions that lead to an error. Roughly speaking, I would like to take these actions into context, but not add these tokens to CE loss.

**Comment by john-b-yang** [comment]
Hi @Martovark! SWE-agent-LM-32B was trained using full (not low rank) fine-tuning. [This script](https://github.com/SWE-bench/SWE-smith/blob/main/swesmith/train/run/ft_torchtune.py) was the one that was actually used.

This sounds interesting, although not quite sure what you mean - are you suggesting modification(s) to [SWE-bench/SWE-smith-trajectories](https://huggingface.co/datasets/SWE-bench/SWE-smith-trajectories), and then training on that?

Could be interesting! Feel free to join [SWE-bench slack](https://join.slack.com/t/swe-bench/shared_invite/zt-36pj9bu5s-o3_yXPZbaH2wVnxnss1EkQ), there's several ongoing conversations about training on SWE-smith 😄

**Comment by Martovark** [comment]
Hi @john-b-yang!
I understand about the full fine-tuning (not LoRA).

Let me try to explain the idea better. Let's look at a dialogue from SWE-smith-traj (as a rule, it has many occurrences of the assistant's answers):
```json
{"role": "system", "content": "sys"},
{"role": "user", "content": "user_1"},
{"role": "assistant", "content": "assistant_1"}, <- "taken into account when calculating the loss"
...
{"role": "user", "content": "user_n"},
{"role": "assistant", "content": "assistant_n"}, <- "taken into account when calculating the loss"
...
```
In the fine-tuning code you linked above, `each` answer from the assistant is taken into account when calculating the loss (each token from `assistant_n` is explicitly involved in calculating the CE loss). 

I suggest not to take some bad answers into account when calculating the loss, for example, when trying to change the code, the assistant can do it incorrectly `several times` and only insert the correct code `at the end`. In such cases, let's take into account only the last insertion of the code in the loss and then the model will learn to do wrong actions much less. 

There is a section in SWE-smith paper: `SWE-agent-LM-32B frequently shows highly repetitive actions
for unresolved instances`. I think these things are related :)

**Comment by Martovark** [comment]
Thats right.
Close agent attempts can be detected with sbert like models. And leave only the last action from the chain of very close ones (cos_sim > 0.99 for example).
